torch_xla/csrc/ir.h

```diff
@@ -110,18 +110,29 @@ class XlaNode : public torch::lazy::Node {
 
   torch::lazy::hash_t node_hash() const { return node_hash_; }
 
-  torch::lazy::hash_t hash() const override { return dag_hash_; }
+  torch::lazy::hash_t hash() const override {
+    if (sharding_hash_ != 0) {
+      return torch::lazy::HashCombine(dag_hash_, sharding_hash_);
+    }
+    return dag_hash_;
+  }
 
   torch::lazy::hash_t shapeHash() const override { return dag_hash_; }
+
+  torch::lazy::hash_t shardingHash() const { return sharding_hash_; }
+
   // The node's outputs get assigned the same HLO sharding
   // TODO: test multi-output example.
   const std::shared_ptr<xla::OpSharding> GetSharding() const {
     return output_sharding_;
   }
-  void SetSharding(const xla::OpSharding& sharding) {
-    output_sharding_ = std::make_shared<xla::OpSharding>(sharding);
+
+  void SetSharding(const xla::OpSharding& sharding);
+
+  void ClearSharding() {
```

Comment:
For a future reference, #4288 , we may need to create a new IR when we clear sharding; the new IR node should return the same hash as the original pre-sharded IR node. cc @JackCaoG

Reply:
Can we implement and test this in #4288 ?